src/libraries/System.Private.CoreLib/src/System/Number.Parsing.cs

```diff
@@ -381,8 +395,24 @@ private static unsafe bool TryParseNumber(ref char* str, char* strEnd, NumberSty
             }
 
             bool negExp = false;
-            number.DigitsCount = digEnd;
+            if (digEnd == maxDigCount && numberOfTrailingZeros > 0 && number.HasNonZeroTail)
+            {
+                // We have an additional non-zero digit in the input past maxDigitCount. We already handle this properly.
+                number.DigitsCount = digEnd;
+            }
+            else
+            {
+                // We have only 0s in the fractional part. We can strip it all out
+                number.DigitsCount = digEnd - numberOfTrailingZeros;
+            }
             number.Digits[digEnd] = (byte)('\0');
+            if (numberOfTrailingZeros != 0)
+            {
+                for (int i = number.DigitsCount; i < digEnd; i++)
+                {
+                    number.Digits[i] = (byte)('\0');
+                }
+            }
```

Comment:
This is going to be "expensive" for production code. I think we can just insert `null` at the first correct position.

We already have what could be considered "garbage" data in the rest of the span due to `SkipLocalsInit` and so we rely on `\0` correctly marking the end.

Reply:
Ah ok. This was an attempt to keep the span clean to avoid future bugs. Cool, I'll set the appropriate digit here.

Reply:
We should largely be doing that in `DEBUG` builds when the span is created: https://source.dot.net/#System.Private.CoreLib/Number.NumberBuffer.cs,44

If you wanted to preserve that here, then you could do a fill of the digits to `0xCC` in DEBUG builds only